reference/run: Clarify the use of numeric UIDs

kblin opened the pull request:
it is possible to pass an UID that has not been created inside the container, clarify this in the docs.
This should fix issue #14795

Signed-off-by: Kai Blin kai@samba.org

kblin committed:
reference/run: Clarify the use of numeric UIDs

it is possible to pass an UID that has not been created inside the container, clarify this in the docs.
This should fix issue #14795

Signed-off-by: Kai Blin <kai@samba.org>

phemmer commented on docs/reference/run.md:
```diff
@@ -1142,10 +1142,11 @@ volume mounted on the host).
-Dockerfile `USER` instruction, but the operator can override it:
+The default user within a container is `root` (id = 0), but if the developer
+created additional users, those are accessible by name.  When passing a numeric
+ID, the user doesn't have to exist in the container. The developer can set a
```
I feel like there should be a paragraph break before `The developer can set a`.

kblin committed:
reference/run: Add a paragraph break

Signed-off-by: Kai Blin <kai@samba.org>

kblin committed:
reference/run: Add some articles

Signed-off-by: Kai Blin <kai@samba.org>

moxiegirl commented on docs/reference/run.md:
```diff
@@ -1142,14 +1142,16 @@ volume mounted on the host).
-The default user within a container is `root` (id = 0), but if the
-developer created additional users, those are accessible too. The
-developer can set a default user to run the first process with the
+The default user within a container is `root` (id = 0), but if the developer
```
@kblin Thank you for the contribution. I think the conten can be clarified a bit more.  Also, can you expand on what you mean by the "user doesn't have to exist in the container" --- is this the `USER` or the `-u` value?  If I pass a user that doesn't exist in the container is the user used anyway.

---> https://gist.github.com/moxiegirl/3cb01145b2fd73f7499d

The  `root` (id = 0) is the default user within a container. An image developer
can create additional users. Those users are accessible by name.  When passing a numeric
user ID, the user doesn't have to exist in the container.

A developer can set a default user to run container processes using the
Dockerfile `USER` instruction. When you start a conatiner, you can override the
`USER` instruction by passing the `-u` option.

    -u="": Username or UID

> **Note:** If you pass a numeric uid, it must be in the `0-2147483647` range.

kblin commented:
@moxiegirl I'm sure it works with -u, see my initial description on issue #14795. I haven't tried USER, but I also fail to see the use case for this, to be honest. As the author of the Dockerfile, you could always just add the user you want to be using later.

I'm not sure I understand your other question. If you pass a user name that doesn't exist, I don't see how the kernel would know which uid to use, so that's probably not what you meant?

moxiegirl commented:
@kblin Thank you, your second paragraph answered my question.

Can you make the editorial changes I provided in the gist?

kblin commented:
Sorry, I didn't realize what the exact goal of the gist was, I'm very used to diffs and missed the difference at the first glance.

I'd change the first sentence to

`root` (id = 0) is the default user in within a container.

so dropping the article. "The root" sounds wrong.

Also, I'd like to stick with "the operator", as that's in line with the rest of the file. If you want to change that overall, that sounds like a separate patch set, and frankly, one I don't care too much about.

kblin committed:
reference/run: Some editorial changes

Based on the suggestions from @moxiegirl, some changes to make the wording more clear.

Signed-off-by: Kai Blin <kai@samba.org>

kblin commented:
I've removed the article in front of `root`, that felt a bit unnatural. I've also stuck with the "operator" nomenclature instead of "you", as that's what the rest of the file uses as well.

thaJeztah commented:
sgtm, so LGTM. Thanks!

moxiegirl commented:
LGTM @kblin thank you for contributing and your patience.
